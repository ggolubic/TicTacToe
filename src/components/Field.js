import React from "react";
import styled, { css } from "styled-components";

export const Field = props => {
  return (
    <FieldDiv
      player={props.player}
      onClick={props.handleTurn}
      content={props.content}
    >
      {props.children}
    </FieldDiv>
  );
};

const FieldDiv = styled.div`
  background-color: ${props => {
    if (props.content === null) return "#14bdac";
    else if (props.content === "x") return "#545454";
    else return "#ffd02d";
  }};
  margin: 5px;
  height: 100px;
  width: 100px;
  border-radius: 10%;
  font-size: 5rem;
  text-align: center;
  line-height: 75px;
  ${props =>
    !props.content &&
    css`
      &:hover {
        background: ${props => (props.player === "x" ? "#545454" : "#ffd02d")};
        transition: 0.3s;
        cursor: pointer;
      }
    `}
`;
