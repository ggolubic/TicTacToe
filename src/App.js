import React from "react";
import styled from "styled-components";

import { Field } from "./components/Field";
import { Score } from "./components/Score";
import Button from "./components/Button";
import Modal from "./components/Modal";

class App extends React.Component {
  state = {
    board: Array(9).fill(null),
    player: "x",
    winner: null,
    score: { x: 0, o: 0, draw: 0 },
    turn: 0
  };

  handleTurn = index => {
    let newBoard = this.state.board;
    if (newBoard[index] === null && !this.state.winner) {
      newBoard[index] = this.state.player;
      this.setState(prevState => ({
        board: newBoard,
        player: this.state.player === "x" ? "o" : "x",
        turn: prevState.turn + 1
      }));
      this.checkWinner();
    }
  };

  checkWinner() {
    let winCombos = [
      ["0", "1", "2"],
      ["3", "4", "5"],
      ["6", "7", "8"],
      ["0", "3", "6"],
      ["1", "4", "7"],
      ["2", "5", "8"],
      ["0", "4", "8"],
      ["2", "4", "6"]
    ];
    this.checkMatch(winCombos);
  }

  checkMatch(winCombos) {
    for (let index = 0; index < winCombos.length; index++) {
      const [a, b, c] = winCombos[index];
      let board = this.state.board;
      if (board[a] && board[a] === board[b] && board[a] === board[c]) {
        let score = this.increaseScore();
        this.setState({
          winner: this.state.player,
          score: score
        });
      }
    }
    if (this.state.turn === 8) {
      let newScore = { ...this.state.score };
      newScore.draw += 1;
      this.setState({ score: newScore });
    }
  }

  increaseScore = () => {
    let newScore = { ...this.state.score };
    newScore[this.state.player]++;
    return newScore;
  };

  newGame = () => {
    const board = Array(9).fill(null);
    this.setState(prevState => ({
      board: board,
      player: prevState.winner === "x" ? "o" : "x",
      winner: null,
      turn: 0
    }));
  };

  resetScore = () => {
    const board = Array(9).fill(null);
    this.setState({
      board: board,
      player: "x",
      winner: null,
      score: { x: 0, o: 0, draw: 0 },
      turn: 0
    });
  };

  render() {
    const { board, player, score, winner } = this.state;
    return (
      <>
        {winner && (
          <Modal
            score={score}
            newGame={this.newGame}
            resetScore={this.resetScore}
          />
        )}
        <Title>Tic Tac Toe App</Title>
        <Score score={score} />
        <Board>
          {board.map((field, index) => (
            <Field
              key={index}
              player={player}
              handleTurn={() => this.handleTurn(index)}
              content={field}
            >
              {field}
            </Field>
          ))}
        </Board>
        <NewGameArea>
          <Button new onClick={this.newGame}>
            New Round
          </Button>
          <Button onClick={this.resetScore}>Reset Score</Button>
        </NewGameArea>
      </>
    );
  }
}

export default App;

const Board = styled.div`
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  display: flex;
  flex-wrap: wrap;
  width: 330px;
`;

export const Title = styled.h1`
  font-family: "Montserrat", sans-serif;
  text-align: center;
`;

export const NewGameArea = styled.div`
  text-align: center;
`;
