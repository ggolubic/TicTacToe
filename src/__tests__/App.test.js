import React from "react";
import Enzyme, { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import "jest-styled-components";

import App from "../App";
import Field from "../components/Field";

Enzyme.configure({ adapter: new Adapter() });

describe("App component", () => {
  it("Should render correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render x*x fields if board size is x", () => {
    const wrapper = mount(<App />);
    expect(wrapper.state().size).toBe(5);
    expect(wrapper.children().find(Field).length).toBe(25);
  });

  it("handleTurn should update board,increase turn and switch player if board position is empty", () => {
    const wrapper = mount(<App />);
    const instance = wrapper.instance();
    expect(wrapper.state("turn")).toBe(0);
    expect(wrapper.state("player")).toBe("x");
    expect(wrapper.state().board[0][0]).toBe("");
    instance.handleTurn(0, 0);
    expect(wrapper.state("turn")).toBe(1);
    expect(wrapper.state("player")).toBe("o");
    expect(wrapper.state().board[0][0]).toBe("x");
  });
  it("handleTurn should do nothing when board position is not empty", () => {
    const wrapper = mount(<App />);
    const instance = wrapper.instance();

    //set content into board
    instance.handleTurn(0, 0);
    expect(wrapper.state("turn")).toBe(1);
    expect(wrapper.state("player")).toBe("o");
    expect(wrapper.state().board[0][0]).toBe("x");

    //try to set content again, state shouldn't change
    instance.handleTurn(0, 0);
    expect(wrapper.state("turn")).toBe(1);
    expect(wrapper.state("player")).toBe("o");
    expect(wrapper.state().board[0][0]).toBe("x");
  });

  it("checkWinner should return false if turn is not high enough", () => {
    const wrapper = shallow(<App />);
    const instance = wrapper.instance();
    expect(instance.checkWinner("x", 0, 0)).toEqual(false);
  });

  it("checkWinner should return draw if board is full and there is no winner", () => {
    const wrapper = shallow(<App />);
    const instance = wrapper.instance();
    wrapper.setState({ size: 3 });
    //init at size 3
    instance.initBoard();
    //fill board
    for (let i = 0; i < 3; i++) {
      for (let j = 0; j < 3; j++) {
        instance.handleTurn(i, j);
      }
    }
    expect(wrapper.state().score.draw).toBe(1);
    expect(wrapper.state("gameDone")).toBe(true);
  });

  it("setWinner should increase score and set gameDone", () => {
    const wrapper = shallow(<App />);
    const instance = wrapper.instance();
    expect(wrapper.state().score.x).toBe(0);
    instance.setWinner();
    expect(wrapper.state().score.x).toBe(1);
    expect(wrapper.state().gameDone).toBe(true);
  });

  it("Should call newGame after clicking on button", () => {
    const wrapper = mount(<App />);
    const instance = wrapper.instance();
    //add something on board
    instance.handleTurn(0, 0);
    expect(wrapper.state().board[0][0]).toBe("x");
    //trigger newGame
    wrapper
      .find("button")
      .first()
      .props()
      .onClick();
    //board should be empty
    expect(wrapper.state().board[0][0]).toBe("");
  });

  it("Should call resetScore after clicking on button", () => {
    const wrapper = mount(<App />);
    const instance = wrapper.instance();
    //add something on board
    instance.handleTurn(0, 0);
    //add winner
    instance.setWinner();
    expect(wrapper.state().board[0][0]).toBe("x");
    //trigger resetScore
    wrapper
      .find("button")
      .last()
      .props()
      .onClick();
    //board should be empty
    expect(wrapper.state().board[0][0]).toBe("");
    //winner should be null
    expect(wrapper.state().winner).toBe(null);
  });
});
