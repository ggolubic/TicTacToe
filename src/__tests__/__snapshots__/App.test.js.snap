// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should render correctly 1`] = `
Array [
  .c0 {
  font-family: "Montserrat",sans-serif;
  text-align: center;
}

<h1
    className="c0"
  >
    Tic Tac Toe App
  </h1>,
  .c1 {
  color: #545454;
}

.c2 {
  color: #ffd02d;
}

.c0 {
  font-family: "Montserrat",sans-serif;
  width: 300px;
  margin: 0 auto;
  text-align: center;
  margin-top: 50px;
}

<div
    className="c0"
  >
    <h2
      className="c1"
    >
      Player 
      1
      : 
      0
    </h2>
    <h2
      className="c2"
    >
      Player 
      2
      : 
      0
    </h2>
    <h2>
      Draw: 
      0
    </h2>
  </div>,
  .c1 {
  background-color: #14bdac;
  margin: 5px;
  position: relative;
  width: calc(100% / 5 - 10px);
  height: calc(100% / 5 - 15px);
  border-radius: 10%;
  text-align: center;
}

.c1:hover {
  background: #545454;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  cursor: pointer;
}

.c1 h2 {
  position: absolute;
  top: 30%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  margin: 0 auto;
  padding: 0;
  font-size: 5vw;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  text-align: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50vw;
  height: 50vw;
}

@media (max-width:1000px) {
  .c1 h2 {
    top: 40%;
  }
}

@media (min-width:1000px) {
  .c0 {
    width: 40vw;
    height: 30vw;
    -webkit-transition: 0.3s;
    transition: 0.3s;
  }
}

<div
    className="c0"
  >
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
    <div
      className="c1"
      content=""
      onClick={[Function]}
      size={5}
    >
      <h2>
        
      </h2>
    </div>
  </div>,
  .c1 {
  font-family: "Montserrat",sans-serif;
  border-radius: 4%;
  background: transparent;
  border: 2px solid #ddbc95;
  color: #ddbc95;
  margin: 0 1rem;
  font-size: 1rem;
  font-weight: 700;
  padding: 0.5rem 1rem;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  cursor: pointer;
  outline: none;
  background: #ddbc95;
  color: white;
}

.c1:hover {
  color: white;
  background: #ddbc95;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c1:hover {
  color: #ddbc95;
  background: white;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c2 {
  font-family: "Montserrat",sans-serif;
  border-radius: 4%;
  background: transparent;
  border: 2px solid #ddbc95;
  color: #ddbc95;
  margin: 0 1rem;
  font-size: 1rem;
  font-weight: 700;
  padding: 0.5rem 1rem;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  cursor: pointer;
  outline: none;
}

.c2:hover {
  color: white;
  background: #ddbc95;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0 {
  text-align: center;
  margin-top: 50px;
}

<div
    className="c0"
  >
    <button
      className="c1"
      onClick={[Function]}
    >
      New Round
    </button>
    <button
      className="c2"
      onClick={[Function]}
    >
      Reset Score
    </button>
  </div>,
]
`;
